sympy/solvers/solvers.py

```diff
@@ -881,11 +877,20 @@ def _sympified_list(w):
                         flags['dict'] = True  # better show symbols
                         symbols.add(p)
                         pot.skip()  # don't go any deeper
-        symbols = list(symbols)
-
         ordered_symbols = False
-    elif len(symbols) == 1 and iterable(symbols[0]):
-        symbols = symbols[0]
+        symbols = list(ordered(symbols))  # to make it canonical
+    else:
+        if len(symbols) == 1 and iterable(symbols[0]):
+            symbols = symbols[0]
+        ordered_symbols = symbols and is_sequence(symbols,
+                        include=GeneratorType)
+        _symbols = list(uniq(symbols))
+        if len(_symbols) != len(symbols):
+            ordered_symbols = False
+            symbols = list(ordered(symbols))
+        else:
+            symbols = _symbols
```

Comment:
```suggestion
        ordered_symbols = False
        if symbols:
            if is_sequence(symbols, include=GeneratorType):
                # the symbols are well ordered; keep the
                # order but make sure they are unique
                _symbols = list(uniq(symbols))
                if len(_symbols) == len(symbols):
                    # the symbols are unique so we can return
                    # solutions relative to their given order
                    ordered_symbols = True
                symbols = _symbols
            else:
                # put them in order for canonical processing
                symbols = list(ordered(symbols))
                if flags.get('set', False):
                    # since symbols were provided, all will be
                    # included in the set output; even though they
                    # were not well ordered, they are now
                    ordered_symbols = True
```